### PR TITLE
[CRB-194][CRB-42] - Publishing event handling

### PR DESCRIPTION
**Pull request by CAGS295**

### CODE REVIEW

File: src/block/block_consensus.rs
Comment (nathanwhit) [inline src/block/block_consensus.rs]: In general, I think `unwrap_err()` in the function body to assert an error is cleaner than using `should_panic`.

File: src/block/block_header.rs
Comment (nathanwhit) [inline src/block/block_header.rs]: What's wrong with the original version that uses `?` here?

File: src/block/paired_fork.rs
Comment (nathanwhit) [inline src/block/paired_fork.rs]: Typo:  `netx` -> `next`

File: src/engine.rs
Comment (nathanwhit) [inline src/engine.rs]: Why a multi-threaded runtime with 1 worker thread instead of just a [single threaded runtime](https://docs.rs/tokio/1.11.0/tokio/runtime/struct.Builder.html#method.new_current_thread)?

**Review by seandeneen-gluwa — DISMISSED** [review]

**Review by atodorov — COMMENTED** [review]
I see there are some pre-existing tests but I don't see them enabled in GitHub Actions, nor I see coverage reports being generated. Should I push a commit to the same branch or to a separate one ?

File: src/futures/update_stream.rs
Comment (nathanwhit) [inline src/futures/update_stream.rs]: If these tests require the `test-futures` flag, why not do 
```suggestion
#[cfg(all(test, feature = "test-futures"))]
```
then you can remove the `cfg_attr` on each test.

File: src/futures/update_stream.rs
Comment (nathanwhit) [inline src/futures/update_stream.rs]:
```suggestion
    match self.updates.try_recv() {
      Ok(update) => {
        trace!("Incoming update {:?}", update);
        match self.node.handle_update(update) {
          Ok(true) => EventResult::Continue,
          Ok(false) => EventResult::Shutdown,
          Err(error) => {
            error!("Update Error: {}", error);
            EventResult::Continue
          }
        }
      }
      Err(std::sync::mpsc::TryRecvError::Disconnected) => {
        error!("Disconnected from validator");
        EventResult::Shutdown
      }
      Err(std::sync::mpsc::TryRecvError::Empty) => {
        sleep(Duration::from_millis(10)).await;
        EventResult::Continue
      }
    }
```
Instead of using the blocking call `recv_timeout` with a timeout of `0`, we can use the `try_recv` method which is intended to be non-blocking.

File: src/miner/error.rs
Comment (nathanwhit) [inline src/miner/error.rs]: We could remove this entirely if we make `Worker::new` infallible. It's also more correct, because an error can never be returned from that function.

File: src/miner/miner.rs
Comment (nathanwhit) [inline src/miner/miner.rs]: See other comment, this could also be removed if we fix `Worker::new`

File: src/node/service.rs
Comment (nathanwhit) [inline src/node/service.rs]: Do we want to panic in this case? Why not just log it and return the error early?

**Review by nathanwhit — CHANGES_REQUESTED** [review]
So my main concern is with the publishing logic, particularly enforcing the block time by using timers. From my reading, we only try to publish once the `PublishSchedulerFuture` signals that we have waited at least `node.config.seconds_between_blocks`. If someone were to, for instance, fork this repo and get rid of that requirement, instead publishing as soon as possible, they would have an advantage in mining, right?

File: src/engine.rs
Comment (CAGS295) [reply src/engine.rs]: The worker parameter is open to discussion, we will change it to a single-threaded runtime if we choose to stick with one thread. Although, if we want to use more threads we may need to split the umbrella future to allow futures to work in parallel.

File: src/node/service.rs
Comment (CAGS295) [reply src/node/service.rs]: ~~We don't want it to panic, but I want the consensus to crash until we know how to handle it. It is not trivial how to handle a missing block, remember that should happen at validation which the validator does, not the consensus. failing to fetch a block at consensus should be equivalent to doing so in the old python PoW impl. Essentially falling the fork. I know I'm leaving all the important questions out. My goal is to flesh the consensus first so that we can focus on the tough questions all at once with a better understanding.~~

**Review by CAGS295 — COMMENTED** [review]
Big PR. Thank you for the input.

**Comment by CAGS295** [comment]
> So my main concern is with the publishing logic, particularly enforcing the block time by using timers. From my reading, we only try to publish once the `PublishSchedulerFuture` signals that we have waited at least `node.config.seconds_between_blocks`. If someone were to, for instance, fork this repo and get rid of that requirement, instead of publishing as soon as possible, they would have an advantage in mining, right?

Yes, but I think that is also possible in sawtooth 1.7. `min_difficulty` should somewhat help. However, how do we encourage miners to wait for the right amount of time? As we discussed earlier the difficulty impl encourages quantity over quality. To eliminate the advantage of maliciously early-publishing and maybe the need for the secondary solver we could make the publisher attempt to publish if a new chain_head is committed instead of discarding the candidate block. This way we could reduce the advantage of the first-on-clock node having a publishing advantage. If I'm not mistaken this is what the secondary solver does implicitly; reducing the advantage of the node that mined the previous head. 

Either way in my mind, this will come at another PR. Remember that this is the first take on a working consensus.

File: src/node/service.rs
Comment (CAGS295) [reply src/node/service.rs]: On further reading, I get what you mean now. I agree.

**Review by nathanwhit — DISMISSED** [review]
> Either way in my mind, this will come at another PR. Remember that this is the first take on a working consensus.

Good point, we can leave that to follow-up PR(s).

This looks good to me, and should be ready to merge (pending #5).

**Review by atodorov — DISMISSED** [review]
@CAGS295 do you mind rebasing so we can see the CI jobs executing ?

**Comment by nathanwhit** [comment]
Updating the lockfile (i.e. `cargo update`) should fix CI.

**Review by atodorov — DISMISSED** [review]
I think you should merge #4 first and then see if anything is left from this PR. Probably not.

**Review by atodorov — DISMISSED** [review]

**Review by atodorov — APPROVED** [review]
@CAGS295 , @nathanwhit is there anything else left to do for this PR ? Can it be merged ?

**Review by nathanwhit — APPROVED** [review]

**Comment by nathanwhit** [comment]
> 
> 
> @CAGS295 , @nathanwhit is there anything else left to do for this PR ? Can it be merged ?

I think this is ready for merging.

**Pull request merged**
